lib/ibm/cloud/sdk/vpc/instances.rb

```diff
@@ -38,21 +38,115 @@ def params(start: nil, limit: nil, name: nil, vpc_id: nil, vpc_crn: nil, vpc_nam
+          end
+
+          # The name of this VM.
+          def name
+            @data[:name]
+          end
+
+          # The status of the virtual server instance. Possible values: [failed,paused,pausing,pending,restarting,resuming,running,starting,stopped,stopping]
+          def status
+            @data[:status]
+          end
+
+          # Whether the state of the VM is in failed state.
+          # @return [Boolean]
+          def failed?
+            status == ERROR_STATE
+          end
+
+          # Whether the state of the VM is in the started state.
+          # @return [Boolean]
+          def started?
+            status == RUNNING_STATE
+          end
+
+          # Whether the state of the VM is in a stopped or paused state.
+          # @return [Boolean]
+          def stopped?
+            STOPPED_STATES.include?(status)
+          end
+
+          # Whether the state of the VM is in a transitional state.
+          # @return [Boolean]
+          def transitional?
+            TRANSITIONAL_STATES.include?(status)
+          end
+
+          # Interact with instance actions.
+          # @return [INSTANCE::Actions]
           def actions
             INSTANCE::Actions.new(self)
           end
 
+          # Interact with instance network interfaces.
+          # @return [INSTANCE::NetworkInterfaces]
           def network_interfaces
             INSTANCE::NetworkInterfaces.new(self)
           end
 
+          # Interact with instance volume attachements.
+          # @return [INSTANCE::VolumeAttachments]
           def volume_attachments
             INSTANCE::VolumeAttachments.new(self)
           end
 
+          # Return the data used for initializing this VM.
           def initialization
             adhoc(method: 'get', path: 'initialization').json
           end
+
+          # Wait for the VM instance to be in a stable state.
+          # @param started [Boolean] When true wait for started, when false wait for stopped.
+          # @param sleep_time [Integer] The time to sleep between refreshes.
+          # @param timeout [Integer] The number of seconds before raising an error.
+          # @param block [Proc] A block to test against. Must return a boolean.
+          # @return [Boolean] The return of the status check.
+          # @raise [RuntimeError] Instance goes into failed state.
+          # @raise [RuntimeError] Timeout has been reached.
+          def wait_for(started: true, sleep_time: 5, timeout: 600, &block)
+            @logger.info("Starting wait for, instance #{id} starts in state #{status}.")
+            until test_state(started: started, &block)
+              raise "VM #{id} is in a failed state." if failed?
+
+              timeout = sleep_counter(sleep_time, timeout)
+              raise "Time out while waiting #{id} to be stable." if timeout <= 0
+            end
+            @logger.info("Finished wait for instance #{id} ends in state #{status}.")
+            test_state(&block)
```

Comment:
Is the point of this last block to return if the wait was successful or not?  Seems like it will either be successful or it'll raise a timeout.
Also if we don't pass `started` and we weren't called with a block wouldn't a call to `wait_for(started: false)` end up calling `test_state` with `started: true`?

Reply:
Uggg, yeah, not including started is a bug.

Reply:
The point of the last is to have a boolean return. Originally I was returning false when an error was hit, but that didn't seem like enough information to determine why so I raised. It probably could have no return since anything that isn't true will raise, and thus the return is useless. I'll remove the last call.

Reply:
Yeah I like the return on success raise on any error its a pretty common pattern.  Typically it is denoted by `!` in the method name so `wait_for!(&:something)` raises an exception and `wait_for(&:something)` returns true/false.
Don't think we need both version but we could name this `wait_for!` so it is obvious that it'll raise on a failure

Reply:
Hah, we could have both versions if we returned a tuple with (boolean, reason) in the standard version and then acted accordingly in the ! version. But that seems like overkill for the moment.